File: others/1300-1400/1353. 最多可以参加的会议数目.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

MOD = int(1e9 + 7)
DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]


class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        mx = max(e[1] for e in events)

        groups = [[] for _ in range(mx + 1)]
        for e in events:
            groups[e[0]].append(e[1])

        ans = 0
        h = []
        for i, g in enumerate(groups):
            while h and h[0] < i:
                heappop(h)
            for end_day in g:
                heappush(h, end_day)
            if h:
                ans += 1
                heappop(h)
        return ans
    

File: others/1300-1400/1353. 最多可以参加的会议数目.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxEvents(vector<vector<int>> &events) {
        ranges::sort(events);
        priority_queue<int, vector<int>, greater<int>> pq;
        int maxDay = 0;
        for (int i = 0; i < events.size(); i++) {
            maxDay = max(maxDay, events[i][1]);
        }
        int ans = 0;
        int j = 0;
        int n = events.size();
        for (int i = events[0][0]; i <= maxDay; ++i) {
            // 把今天开始的活动加入队列
            while (j < n && events[j][0] <= i) {
                pq.emplace(events[j][1]);
                ++j;
            }
            // 移除所有已经过期的活动
            while (!pq.empty() && pq.top() < i) {
                pq.pop();
            }
            // 可以参加一个活动
            if (!pq.empty()) {
                pq.pop();
                ++ans;
            }
        }
        return ans;
    }
};